### PR TITLE
VRTConverter and XVSVesting Contracts

### PR DESCRIPTION
**Pull request by defistar**
**Requirement:**

- VRTConverter: Users can convert the VRT to XVS

- XVSVesting: XVS Tokens to redeemed by VRT conversion are vested in XVSVesting Contract.
Users can have multiple XVS Vesting Actions and each vesting is linearly vested over a period of 360 days
XVSVesting contract is funded with XVS to ensure users can withdraw/claim the vested XVS

- [ ] I have updated the documentation to account for the smartContract functions
- [x] If I added new functionality, I added tests covering it.

### CODE REVIEW

File: contracts/VRT/XVSVesting.sol [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+            vesting.amount,
+            vesting.withdrawnAmount
+        );
+    }
+
+    function withdraw() external vestingExistCheck(msg.sender) {
+        require(
+            vrtConversionAddress != address(0),
+            "VRT-Conversion Address is not set"
+        );
```
[the remark this answers is not in the document]
Comment (defistar) [reply contracts/VRT/XVSVesting.sol:175]: @kkirka  will remove.

File: contracts/VRT/XVSVesting.sol [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+
+        VestingRecord memory vesting = VestingRecord({
+            recipient: recipient,
+            startTime: getCurrentTime(),
+            amount: depositAmount,
+            withdrawnAmount: 0
+        });
+
+        vestingsOfRecipient.push(vesting);
+        numberOfVestings[recipient] = numberOfVestings[recipient].add(1);
```
[the remark this answers is not in the document]
Comment (kkirka) [reply contracts/VRT/XVSVesting.sol]: Solidity doesn't compute the length, it's already in storage: https://docs.soliditylang.org/en/latest/internals/layout_in_storage.html#mappings-and-dynamic-arrays

image

File: contracts/VRT/XVSVesting.sol [hunk 3]
```diff
@@ -0,0 +1,257 @@
...
+        require(
+            vestings[recipient].length > 0,
+            "recipient doesnot have any vestingRecord"
+        );
+
+        _;
+    }
+
+    function deposit(address recipient, uint depositAmount) external onlyVrtConverter
+        nonReentrant
```
Comment (kkirka) [inline contracts/VRT/XVSVesting.sol:145]: No need for reentrancy checks here, we don't do any transfers

File: contracts/VRT/XVSVesting.sol [hunk 4]
```diff
@@ -0,0 +1,257 @@
...
+        uint256 currentTime = getCurrentTime();
+
+        for(uint i = 0; i < vestingCount; i++) {
+            VestingRecord memory vesting = vestingsOfRecipient[i];
+            uint256 vestedAmount = calculateVestedAmount(vesting.amount, vesting.startTime, currentTime);
+            totalVestedAmount = totalVestedAmount.add(vestedAmount);
+        }
+
+        return totalVestedAmount;
+    }
```
Comment (kkirka) [inline contracts/VRT/XVSVesting.sol:229]: Looks like the logic is the same as in `getWithdrawableAmount`, and it's not called from within the contract (so we don't need to worry much about the gas spendings) – do we need a separate `getVestedAmount` function then?

**Comment by kkirka** [comment]
Task linked: [CU-25cn4dv Change XVS storage](https://app.clickup.com/t/25cn4dv)

**Comment by kkirka** [comment]
Task linked: [CU-25cn3xx Remove daily limit on conversion](https://app.clickup.com/t/25cn3xx)

**Comment by kkirka** [comment]
Task linked: [CU-25cn43p Add view functions to track total amounts](https://app.clickup.com/t/25cn43p)

**Comment by kkirka** [comment]
Task linked: [CU-25cn3ue Allow multiple vesting records](https://app.clickup.com/t/25cn3ue)

File: contracts/VRT/XVSVesting.sol (hunk 2 above)
[the remark this answers is not in the document]
Comment (defistar) [reply contracts/VRT/XVSVesting.sol]: @kkirka removed the storage variable

File: contracts/VRT/XVSVesting.sol (hunk 4 above)
Comment (defistar) [reply contracts/VRT/XVSVesting.sol:229]: @kkirka 
this is for UI/client to display amount vested for an address.
we can make it external

Comment (defistar) [reply contracts/VRT/XVSVesting.sol:228]: @kkirka 
it includes `vestedAmount`. same math as in `getVestedAmount`. 
infact client can use the function `getWithdrawableAmount` 

`getWithdrawableAmount`  - 3 values:  `totalWithdrawableAmount` , `totalVestedAmount` , `totalWithdrawnAmount`
`getVestedAmount` is to get only vestedAmount

Comment (defistar) [reply contracts/VRT/XVSVesting.sol:228]: @kkirka 
shall i remove the `getVestedAmount`
and let client use only `getWithdrawableAmount` and infer the value needed from response ?

**Review by RoseRompkxm — APPROVED** [review]

**Comment by kkirka** [comment]
LGTM

**Pull request merged**